python/cudf/cudf/core/frame.py

```diff
@@ -1232,10 +1232,22 @@ def drop_duplicates(
         return result
 
     def replace(self, to_replace, replacement):
+        if isinstance(self, cudf.MultiIndex):
+            raise NotImplementedError(
+                "MultiIndex not yet supported for replace."
+            )
+
         copy_data = self._data.copy()
 
         for name, col in copy_data.items():
             if not (to_replace is None and replacement is None):
+                # For Index replace if replacement has string change
+                # thecolumn type to string so mixed replacements
+                # will select the proper column type and function.
+                if not hasattr(self, "index") and any(
+                    type(item) == str for item in replacement
+                ):
+                    col = col.astype(str)
```

Comment:
I think this behavior should be consistent regardless of whether it's an Index or not.

@brandon-b-miller @shwina are we sure we want this to be the behavior for `replace` as opposed to throwing? Pandas returns an `object` dtype of mixed types here so our code can't be 100% compatible.

Reply:
I agree that it should be, but without the check for index there are 36 tests that fail.

Reply:
Yea, I believe the tests are failing because we're trying to change the dtype via the `.rename` call. What you've done here is make it allow changing the dtype only if it's an Index as opposed to allowing changing the dtype in general. We should be consistent regardless.

Reply:
To be clear, does this enable doing something like:

```
a = cudf.Series([1, 2, 3])
b = a.replace(1, 'x')
# b is now of type `str`
```

If so, this is not the right behaviour and we should 100% throw instead. We aren't really getting Pandas behaviour here anyway -- Pandas maintains the types of `2` and `3` as `int`.

Reply:
Easiest thing would be to throw. I think honestly the more fundamental issue here is the fact that every datatype that the index could possibly hold has some limits and there are some restrictions on what items you can represent within those restrictions. Here it's obviously a problem since you can't put a string in a homogenous int array. But more generally there's lots of floats for instance we can't just translate over to int64. What we do in this type of situation should be as uniform as possible across the codebase in my opinion. In joins at least, we chose to scan the data and explicitly check if we could safely cast the incoming data to the destination datatype (except in lucky situations like casting int32 to int64).

Reply:
On the other hand throwing might actually get users to think a little bit more carefully about what they're doing now that they're in GPU land which might be useful anyways.

Reply:
@brandon-b-miller is the logic for whether or not to typecast something that can be generalized outside of join to use it here? That sounds like the best approach from my perspective.

Reply:
Right now it lives in join as well as the `can_cast_safely` functions attached to most of our column classes since the rules depend on the type of join in some cases. But in the most common cases it's just `np.find_common_type` in the end, and that's fairly easy to use anywhere.